tdms/tests/unit/array_tests/test_XYZTensor3D.cpp

```diff
@@ -0,0 +1,43 @@
+/**
+ * @file test_XYZTensor3D.cpp
+ * @author William Graham
+ * @brief Tests allocation, deallocation, and methods for the XYZTensor3D class
+ */
+#include <catch2/catch_test_macros.hpp>
+#include <spdlog/spdlog.h>
+
+#include "arrays.h"
+#include "globals.h"
+
+const double tol = 1e-8;
+
+TEST_CASE("Test XYZTensor allocation") {
+  SPDLOG_INFO("== Testing XYZTensor3D allocation/deallocation");
+  // dimensions for the test-tensor
+  int n_layers = 4, n_cols = 8, n_rows = 16;
+
+  // try creating an XYZTensor3D using the default constructor
+  XYZTensor3D<double> default_constructor_test;
+  // check that, although this has been declared, its members still point to nullptrs
+  // we should be able to index this through AxialDirection
+  REQUIRE(default_constructor_test[AxialDirection::X] == nullptr);
+  REQUIRE(default_constructor_test[AxialDirection::Y] == nullptr);
+  REQUIRE(default_constructor_test[AxialDirection::Z] == nullptr);
+
+  // now try to allocate the components
+  default_constructor_test.allocate(n_rows, n_cols, n_layers);
+  // we should be able to safely populate each of the XYZ tensors now
+  for (char component : {'x', 'y', 'z'}) {
+    for (int k = 0; k < n_layers; k++) {
+      for (int j = 0; j < n_cols; j++) {
+        for (int i = 0; i < n_rows; i++) {
+          // shouldn't seg fault
+          REQUIRE_NOTHROW(default_constructor_test[component][k][j][i] = 0.);
+        }
+      }
+    }
+  }
+  // at end, destructor should be called as appropriate
+  // but this class has no destructor --- it doesn't store the size of its arrays
+  // so we should use Tensor3Ds for the components no??
```

Comment:
Right....

> ### XYZTensor3D tests
>
>   * Tests the default constructor
 >  * Tests that the XYZ components can be accessed through the AxialDirection enum
>   * Tests that the XYZ components can be accessed by providing 'x','y','z' chars
>   * Tests that, post-allocation, elements can be assigned and accessed

I now can't find where these guys are explicitly `free`d? If the destructor doesn't exit it will be the default, and presumably won't deallocate all of the stuff we've allocated in `XYZTensor::allocate`. Do we explicitly free it all in `iterator.cpp`?

 `malloc` with no `free` ? Or do I miss something?

Reply:
If we're templating it... perhaps we want to try calling allocate on a `XYZTensor<std::complex>` and check for memory leaks.

Reply:
> `malloc` with no `free` ? Or do I miss something?

As a standalone class, this is a `malloc` with no free since there's no custom destructor. Also, for whatever reason we _don't_ store the dimensions of the XYZTensor3D arrays we allocate...

However, we only actually use `XYZTensor3D` as an attribute of `Field`, which _does_ free the memory we assigned in it's [custom destructor](https://github.com/UCL/TDMS/blob/1dfd74e37ff5c16818921d17393d36c1d79b927c/tdms/src/fields/base.cpp#L62-L70). Although that being said, there's no check in `Field` that this memory has been _assigned_ in the first place either :cry: 

I'll add a custom destructor to `XYZTensor3D` (_and_ check the other classes for potentially similar issues) that
1. Checks whether memory has been assigned
2. If so, free it on destruction
3. Otherwise, silently exit
`Field` will likewise receive an update.